pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,574 @@
+	return false
+}
+
+func (plugin *storageosPlugin) SupportsBulkVolumeVerification() bool {
+	return false
+}
+
+func getVolumeSource(spec *volume.Spec) (*v1.StorageOSVolumeSource, bool, error) {
+	if spec.Volume != nil && spec.Volume.StorageOS != nil {
+		return spec.Volume.StorageOS, spec.Volume.StorageOS.ReadOnly, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.StorageOS != nil {
+		return spec.PersistentVolume.Spec.StorageOS, spec.ReadOnly, nil
+	}
+	return nil, false, fmt.Errorf("Spec does not reference a StorageOS volume type")
+}
+
+// storageosManager is the abstract interface to StorageOS volume ops.
+type storageosManager interface {
+	// Creates a StorageOS volume.
+	CreateVolume(provisioner *storageosProvisioner) (*storageosVolume, error)
+	// Attaches the disk to the kubelet's host machine.
+	AttachVolume(mounter *storageosMounter) (string, error)
+	// Detaches the disk from the kubelet's host machine.
+	DetachVolume(unmounter *storageosUnmounter, dir string) error
+	// Mounts the disk on the Kubelet's host machine.
+	MountVolume(mounter *storageosMounter, mnt, dir string) error
+	// Unmounts the disk from the Kubelet's host machine.
+	UnmountVolume(unounter *storageosUnmounter) error
+	// Deletes the storageos volume.  All data will be lost.
+	DeleteVolume(deleter *storageosDeleter) error
+}
+
+// storageos volumes represent a bare host directory mount of an StorageOS export.
+type storageos struct {
+	podUID       types.UID
+	namespace    string
+	pvName       string
+	volName      string
+	volNamespace string
+	description  string
+	pool         string
+	fsType       string
+	size         int
+	labels       map[string]string
+	manager      storageosManager
+	mounter      mount.Interface
+	plugin       *storageosPlugin
+	volume.MetricsProvider
+}
+
+type storageosMounter struct {
+	*storageos
+	fsType     string
+	readOnly   bool
+	devicePath string
+	// Interface used to mount the file or block device
+	diskMounter *mount.SafeFormatAndMount
+}
+
+var _ volume.Mounter = &storageosMounter{}
+
+func (b *storageosMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        b.readOnly,
+		Managed:         !b.readOnly,
+		SupportsSELinux: true,
+	}
+}
+
+// Checks prior to mount operations to verify that the required components (binaries, etc.)
+// to mount the volume are available on the underlying node.
+// If not, it returns an error
+func (b *storageosMounter) CanMount() error {
+	return nil
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *storageosMounter) SetUp(fsGroup *types.UnixGroupID) error {
+	// Need a namespace to find the volume, try pod's namespace if not set.
+	if b.volNamespace == "" {
+		glog.V(2).Infof("Setting StorageOS volume namespace to pod namespace: %s", b.namespace)
+		b.volNamespace = b.namespace
+	}
+
+	// Attach the StorageOS volume as a block device
+	devicePath, err := b.manager.AttachVolume(b)
```

Comment:
yes - we have a global namespace so all volumes are visible on all nodes without attaching, though only a single node can mount at a time.  The AttachVolume here does some setup if the volume's device can't be used directly.